proto/api.go

```diff
@@ -51,40 +52,179 @@ const (
 	isWrite                // write cmds go through raft and must be proposed on leader
 	isTxnWrite             // txn write cmds start heartbeat and are marked for intent resolution
 	isRange                // range commands may span multiple keys
+	isReverse              // reverse commands traverse ranges in descending direction
 )
 
+/* TODO(tschottdorf): All of those args.(*BatchRequest) branches in
+* the flag methods are going away soon, when BatchRequest stops being a
+* proto.Request. */
+
 // IsAdmin returns true if the request is an admin request.
 func IsAdmin(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsAdmin()
+	}
 	return (args.flags() & isAdmin) != 0
 }
 
 // IsRead returns true if the request is a read request.
 func IsRead(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsRead()
+	}
 	return (args.flags() & isRead) != 0
 }
 
 // IsWrite returns true if the request is a write request.
 func IsWrite(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsWrite()
+	}
 	return (args.flags() & isWrite) != 0
 }
 
 // IsReadOnly returns true if the request is read-only.
 func IsReadOnly(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsReadOnly()
+	}
 	return IsRead(args) && !IsWrite(args)
 }
 
+// IsReverse returns true if the request is reverse.
+func IsReverse(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsReverse()
+	}
+	return (args.flags() & isReverse) != 0
+}
+
 // IsTransactionWrite returns true if the request produces write
 // intents when used within a transaction.
 func IsTransactionWrite(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsTransactionWrite()
+	}
 	return (args.flags() & isTxnWrite) != 0
 }
 
 // IsRange returns true if the operation is range-based and must include
 // a start and an end key.
 func IsRange(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsRange()
+	}
 	return (args.flags() & isRange) != 0
 }
 
+/* TODO(tschottdorf): all of these are easier to implement when
+* taking the union over the flags of each request.
+* TODO(tschottdorf): won't need all of these Is* methods below on batch,
+* at least not here. Just use those that Spencer has in his WIP.
+ */
+
+// IsAdmin returns true iff the BatchRequest contains an admin request.
+func (br *BatchRequest) IsAdmin() bool {
```

Comment:
going to clean this up (all those assertions above, and not using `flags()` here).